insonmnia/npp/puncher.go

```diff
@@ -0,0 +1,302 @@
+	defer cancel()
+
+	return m.DialContext(ctx, addr)
+}
+
+func (m *natPuncher) DialContext(ctx context.Context, addr common.Address) (net.Conn, error) {
+	addrs, err := m.resolve(ctx, addr)
+	if err != nil {
+		return nil, err
+	}
+
+	return m.punch(ctx, addrs)
+}
+
+func (m *natPuncher) Accept() (net.Conn, error) {
+	// Check for acceptor listenerChannel, if there is a connection - return immediately.
+	select {
+	case conn := <-m.listenerChannel:
+		m.log.Info("received acceptor peer", zap.Any("conn", conn))
+		return conn.unwrap()
+	default:
+	}
+
+	addrs, err := m.publish(m.ctx)
+	if err != nil {
+		m.log.Error("failed to publish itself on the rendezvous", zap.Error(err))
+		return nil, TransportError{err}
+	}
+
+	m.log.Info("received remote peer endpoints", zap.Any("addrs", addrs))
+
+	ctx, cancel := context.WithTimeout(m.ctx, m.timeout)
+	defer cancel()
+
+	// Here the race begins! We're simultaneously trying to connect to ALL
+	// provided endpoints with a reasonable timeout. The first winner will
+	// be the champion, while others die in agony. Life is cruel.
+	conn, err := m.punch(ctx, addrs)
+	if err != nil {
+		return nil, err
+	}
+
+	// TODO: At last when there is no hope, use relay server.
+	return conn, nil
+}
+
+func (m *natPuncher) resolve(ctx context.Context, addr common.Address) (*sonm.RendezvousReply, error) {
+	privateAddrs, err := m.privateAddrs()
+	if err != nil {
+		return nil, err
+	}
+
+	request := &sonm.ConnectRequest{
+		Protocol:     protocol,
+		PrivateAddrs: []*sonm.Addr{},
+		ID:           addr.String(),
+	}
+
+	request.PrivateAddrs, err = convertAddrs(privateAddrs)
+	if err != nil {
+		return nil, err
+	}
+
+	return m.client.Resolve(ctx, request)
+}
+
+func (m *natPuncher) publish(ctx context.Context) (*sonm.RendezvousReply, error) {
+	privateAddrs, err := m.privateAddrs()
+	if err != nil {
+		return nil, err
+	}
+
+	request := &sonm.PublishRequest{
+		PrivateAddrs: []*sonm.Addr{},
+	}
+
+	request.PrivateAddrs, err = convertAddrs(privateAddrs)
+	if err != nil {
+		return nil, err
+	}
+
+	return m.client.Publish(ctx, request)
+}
+
+func convertAddrs(addrs []net.Addr) ([]*sonm.Addr, error) {
+	var result []*sonm.Addr
+	for _, addr := range addrs {
+		host, port, err := netutil.SplitHostPort(addr.String())
+		if err != nil {
+			return nil, err
+		}
+
+		result = append(result, &sonm.Addr{
+			Protocol: protocol,
+			Addr: &sonm.SocketAddr{
+				Addr: host.String(),
+				Port: uint32(port),
+			},
+		})
+	}
+
+	return result, nil
+}
+
+func (m *natPuncher) punch(ctx context.Context, addrs *sonm.RendezvousReply) (net.Conn, error) {
+	if addrs.Empty() {
+		return nil, fmt.Errorf("no addresses resolved")
+	}
+
+	m.log.Debug("punching", zap.Any("addrs", addrs))
+
+	pending := make(chan connTuple)
+
+	// An example of clean architecture (TM) of how to gracefully close a listenerChannel.
+	// Nah, just kidding. Fuck go.
+	wg := &sync.WaitGroup{}
+	wg.Add(1 + len(addrs.PrivateAddrs))
+	go func() {
```

Comment:
in general you don't need this. Just read (on line 239) from this channel (1 + len(addrs.PrivateAddrs)) times